compiler/src/main/kotlin/io/github/apl_cornell/viaduct/analysis/InformationFlowAnalysis.kt

```diff
@@ -56,631 +72,622 @@
-            }
-        }
+    // private val solution by lazy { constraintSystem.solution() }
 
-    /**
-     * Name of the PC label at a particular node.
-     */
+    /** Name of the PC label at a particular node. */
     private val Node.pathName: String by attribute {
         val parent = tree.parent(this)
         when {
             parent == null -> "program"
 
             this is FunctionDeclarationNode -> "${parent.pathName}.${name.value.name}"
 
-            parent is IfNode && this is BlockNode -> "${parent.pathName}.if.${tree.childIndex(this)}"
+            this is IfNode -> "${parent.pathName}.${tree.childIndex(this)}.if"
 
-            this is InfiniteLoopNode -> "${parent.pathName}.loop"
+            parent is IfNode && this is BlockNode -> "${parent.pathName}.${tree.childIndex(this)}"
+
+            this is InfiniteLoopNode -> "${parent.pathName}.${tree.childIndex(this)}.loop"
 
             else -> parent.pathName
         }
     }
 
-    /**
-     * Adds a constraint asserting that [node] with label [nodeLabel] can flow to a location
-     * with label [to].
-     */
-    private fun assertFlowsTo(
-        solver: ConstraintSolver<InformationFlowError>,
-        node: HasSourceLocation,
-        nodeLabel: AtomicLabelTerm,
-        to: LabelTerm
+    /** The [LabelTerm] that represents the label of the program counter at this node. */
+    private val Node.pcTerm: LabelTerm
+        get() =
+            when (this) {
+                is FunctionDeclarationNode ->
+                    term(this.pcLabel.value.interpret())
+
+                else -> term(LabelVariable.PC(pathName))
+            }
+
+    /** The [LabelTerm] that represents the label of the [Variable] declared by this node. */
+    // TODO: incorporate label annotations
+    private val VariableDeclarationNode.labelTerm: LabelTerm
+        get() =
+            when (this) {
+                is ParameterNode -> {
+                    term((objectType.labelArguments!!.first().value.interpret()))
+                }
+
+                is LetNode -> {
+                    term(LabelVariable.Data.Declaration(this))
+                }
+
+                is DeclarationNode -> {
+                    if (objectType.labelArguments == null) {
+                        term(LabelVariable.Data.Declaration(this))
+                    } else {
+                        term((objectType.labelArguments.first().value.interpret()))
+                    }
+                }
+
+                is ObjectDeclarationArgumentNode -> {
+                    term(LabelVariable.Data.Declaration(this))
+                }
+            }
+
+    /** Returns constraints asserting that [this] node with the given label can flow to a location with label [to]. */
+    private infix fun Pair<HasSourceLocation, LabelTerm>.flowsTo(toLabel: LabelTerm): Sequence<LabelConstraint> {
+        val constraints: Iterable<LabelConstraint> =
+            second.flowsTo(toLabel, LabelConstant.bounds()) { to, from ->
```

Comment:
Arguments `to` and `from` are flipped in most of these functions. You should double check all of them.